File: flink/flink-train-scala/src/main/java/com/imooc/flink/course04/DataSetTransformationApp.java
package com.imooc.flink.course04;

import com.imooc.flink.scala.course04.DBUtils;
import org.apache.flink.api.common.functions.*;
import org.apache.flink.api.common.operators.Order;
import org.apache.flink.api.java.ExecutionEnvironment;
import org.apache.flink.api.java.operators.DataSource;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.util.Collector;
import scala.Int;

import java.util.ArrayList;
import java.util.List;

public class DataSetTransformationApp {
    public static void main(String[] args) throws Exception {
        ExecutionEnvironment environment = ExecutionEnvironment.getExecutionEnvironment();
        //mapFunction(environment);
        //fileterFunction(environment);
        mapPartitionFunction(environment);
        //firstFunction(environment);
        //flatMapFunction(environment);
        //distinctFunction(environment);
        //joinFunction(environment);
        //outerJoinFunction(environment);
        //crossFunction(environment);
    }

    private static void crossFunction(ExecutionEnvironment environment) throws Exception {
        List<String> info1 = new ArrayList<>();
        info1.add("曼联");
        info1.add("曼城");

        List<Integer> info2 = new ArrayList<>();
        info2.add(3);
        info2.add(1);
        info2.add(0);

        DataSource<String> data1 = environment.fromCollection(info1);
        DataSource<Integer> data2 = environment.fromCollection(info2);

        data1.cross(data2).print();
    }

    private static void outerJoinFunction(ExecutionEnvironment env) throws Exception {
        List<Tuple2<Integer, String>> info1 = new ArrayList<Tuple2<Integer, String>>();
        info1.add(new Tuple2<>(1, "hadoop"));
        info1.add(new Tuple2<>(2, "spark"));
        info1.add(new Tuple2<>(3, "flink"));
        info1.add(new Tuple2<>(4, "java"));
        info1.add(new Tuple2<>(5, "Spring"));
        info1.add(new Tuple2<>(6, "Linux"));

        List<Tuple2<Integer, String>> info2 = new ArrayList<Tuple2<Integer, String>>();
        info2.add(new Tuple2<>(1, "hadoop"));
        info2.add(new Tuple2<>(2, "spark"));
        info2.add(new Tuple2<>(3, "flink"));
        info2.add(new Tuple2<>(4, "java"));
        info2.add(new Tuple2<>(5, "Spring"));
        info2.add(new Tuple2<>(7, "Vue"));

        DataSource<Tuple2<Integer, String>> data1 = env.fromCollection(info1);
        DataSource<Tuple2<Integer, String>> data2 = env.fromCollection(info2);

        data1.leftOuterJoin(data2).where(0).equalTo(0).with(new JoinFunction<Tuple2<Integer, String>, Tuple2<Integer, String>, Tuple3<Integer, String, String>>() {
            @Override
            public Tuple3<Integer, String, String> join(Tuple2<Integer, String> first, Tuple2<Integer, String> second) throws Exception {
                if (second == null) {
                    return new Tuple3<>(first.f0, first.f1, "-");
                } else {
                    return new Tuple3<>(first.f0, first.f1, second.f1);
                }
            }
        }).print();

        System.out.println("==============================");

        data1.rightOuterJoin(data2).where(0).equalTo(0).with(new JoinFunction<Tuple2<Integer, String>, Tuple2<Integer, String>, Tuple3<Integer, String, String>>() {
            @Override
            public Tuple3<Integer, String, String> join(Tuple2<Integer, String> first, Tuple2<Integer, String> second) throws Exception {
                if (first == null) {
                    return new Tuple3<>(second.f0, "-", second.f1);
                } else {
                    return new Tuple3<>(first.f0, first.f1, second.f1);
                }
            }
        }).print();

        System.out.println("===============================");

        data1.fullOuterJoin(data2).where(0).equalTo(0).with(new JoinFunction<Tuple2<Integer, String>, Tuple2<Integer, String>, Tuple3<Integer, String, String>>() {
            @Override
            public Tuple3<Integer, String, String> join(Tuple2<Integer, String> first, Tuple2<Integer, String> second) throws Exception {
                if (first == null) {
                    return new Tuple3<>(second.f0, "-", second.f1);
                } else if (second == null) {
                    return new Tuple3<>(first.f0, first.f1, "-");
                } else {
                    return new Tuple3<>(first.f0, first.f1, second.f1);
                }
            }
        }).print();
    }


    public static void joinFunction(ExecutionEnvironment env) throws Exception {
        List<Tuple2<Integer, String>> info1 = new ArrayList<Tuple2<Integer, String>>();
        info1.add(new Tuple2<>(1, "hadoop"));
        info1.add(new Tuple2<>(2, "spark"));
        info1.add(new Tuple2<>(3, "flink"));
        info1.add(new Tuple2<>(4, "java"));
        info1.add(new Tuple2<>(5, "Spring"));
        info1.add(new Tuple2<>(6, "Linux"));

        List<Tuple2<Integer, String>> info2 = new ArrayList<Tuple2<Integer, String>>();
        info2.add(new Tuple2<>(1, "hadoop"));
        info2.add(new Tuple2<>(2, "spark"));
        info2.add(new Tuple2<>(3, "flink"));
        info2.add(new Tuple2<>(4, "java"));
        info2.add(new Tuple2<>(5, "Spring"));
        info2.add(new Tuple2<>(7, "Vue"));

        DataSource<Tuple2<Integer, String>> data1 = env.fromCollection(info1);
        DataSource<Tuple2<Integer, String>> data2 = env.fromCollection(info2);
        
        data1.join(data2).where(0).equalTo(0).with(new JoinFunction<Tuple2<Integer, String>, Tuple2<Integer, String>, Tuple3<Integer, String, String>>() {
            @Override
            public Tuple3<Integer, String, String> join(Tuple2<Integer, String> first, Tuple2<Integer, String> second) throws Exception {
                return new Tuple3<>(first.f0, first.f1, second.f1);
            }
        }).print();
    }

    public static void distinctFunction(ExecutionEnvironment env) throws Exception {
        List<String> info = new ArrayList<>();
        info.add("hadoop,spark");
        info.add("hadoop, flink");
        info.add("flink,flink");
        DataSource<String> data = env.fromCollection(info);
        data.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public void flatMap(String value, Collector<String> collector) throws Exception {
                String[] splits = value.split(",");
                for (String split : splits) {
                    collector.collect(split.trim());
                }
            }
        }).distinct().print();
    }

    public static void flatMapFunction(ExecutionEnvironment env) throws Exception {
        List<String> info = new ArrayList<>();
        info.add("hadoop,spark");
        info.add("hadoop, flink");
        info.add("flink,flink");
        DataSource<String> data = env.fromCollection(info);
        data.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public void flatMap(String value, Collector<String> collector) throws Exception {
                String[] splits = value.split(",");
                for (String split : splits) {
                    collector.collect(split.trim());
                }
            }
        }).map(new MapFunction<String, Tuple2<String, Integer>>() {
            @Override
            public Tuple2<String, Integer> map(String value) throws Exception {
                return new Tuple2<>(value, 1);
            }
        }).groupBy(0).sum(1).print();
    }

    public static void firstFunction(ExecutionEnvironment env) throws Exception {
        List<Tuple2<Integer, String>> info = new ArrayList<Tuple2<Integer, String>>();
        info.add(new Tuple2<>(1, "hadoop"));
        info.add(new Tuple2<>(1, "spark"));
        info.add(new Tuple2<>(1, "flink"));
        info.add(new Tuple2<>(2, "java"));
        info.add(new Tuple2<>(2, "Spring"));
        info.add(new Tuple2<>(3, "Linux"));
        info.add(new Tuple2<>(4, "Vue"));

        DataSource<Tuple2<Integer, String>> data = env.fromCollection(info);
        data.first(3).print();
        System.out.println("======================");
        data.groupBy(0).first(2).print();
        System.out.println("======================");
        data.groupBy(0).sortGroup(1, Order.DESCENDING).first(2).print();

    }

    public static void mapPartitionFunction(ExecutionEnvironment env) throws Exception {
        List<String> list = new ArrayList<String>();
        for (int i = 0; i <= 100; i++) {
            list.add("student" + i);
        }

        DataSource<String> dataSource = env.fromCollection(list).setParallelism(6);
//        dataSource.map(new MapFunction<String, String>() {
//            @Override
//            public String map(String input) throws Exception {
//                String connection = DBUtils.getConnnection();
//                System.out.println("connection = " + connection);
//                DBUtils.returnConnection(connection);
//
//                return input;
//            }
//        }).print();
        dataSource.mapPartition(new MapPartitionFunction<String, String>() {
            @Override
            public void mapPartition(Iterable<String> inputs, Collector<String> collector) throws Exception {
                String connection = DBUtils.getConnnection();
                System.out.println("connection = " + connection);
                DBUtils.returnConnection(connection);
                for (String word : inputs) {
                    collector.collect(word);
                }
            }
        }).print();
    }


    public static void fileterFunction(ExecutionEnvironment env) throws Exception {
        List<Integer> list = new ArrayList<Integer>();
        for (int i = 0; i <= 10; i++) {
            list.add(i);
        }

        DataSource<Integer> dataSource = env.fromCollection(list);
        dataSource.map(new MapFunction<Integer, Integer>() {
            @Override
            public Integer map(Integer value) throws Exception {
                return value + 1;
            }
        }).filter(new FilterFunction<Integer>() {
            @Override
            public boolean filter(Integer value) throws Exception {
                return value > 5;
            }
        }).print();
    }
    
    public static void mapFunction(ExecutionEnvironment env) throws Exception {
        List<Integer> list = new ArrayList<Integer>();
        for (int i = 0; i <= 10; i++) {
            list.add(i);
        }
        DataSource<Integer> dataSource = env.fromCollection(list);
        dataSource.map(new MapFunction<Integer, Integer>() {
            @Override
            public Integer map(Integer value) throws Exception {
                return value + 1;
            }
        }).print();
    }
}

File: flink/flink-train-scala/src/main/scala/com/imooc/flink/scala/course04/DataSetTransformationApp.scala
package com.imooc.flink.scala.course04

import org.apache.flink.api.common.operators.Order
import org.apache.flink.api.scala.ExecutionEnvironment

import scala.collection.mutable.ListBuffer

object DataSetTransformationApp {
    def main(args: Array[String]): Unit = {
        val env = ExecutionEnvironment.getExecutionEnvironment
        //mapFunction(env)
        //filterFunction(env)
        //mapPartitionFunction(env)
        //firstFunction(env)
        //flatMapFunction(env)
        //distinctFunction(env)
        //joinFunction(env)
        outerJoinFunction(env)
    }

    def outerJoinFunction(env: ExecutionEnvironment): Unit ={
        val info1 = ListBuffer[(Int, String)]()
        info1.append((1, "hadoop"))
        info1.append((2, "spark"))
        info1.append((3, "flink"))
        info1.append((4, "java"))
        info1.append((5, "Spring"))
        info1.append((6, "Linux"))

        val info2 = ListBuffer[(Int, String)]()
        info2.append((1, "beijing"))
        info2.append((2, "shanghai"))
        info2.append((3, "chengdu"))
        info2.append((4, "guangzhou"))
        info2.append((5, "hangzhou"))
        info2.append((7, "shenzhen"))

        import org.apache.flink.api.scala._
        val data1 = env.fromCollection(info1)
        val data2 = env.fromCollection(info2)

        data1.leftOuterJoin(data2).where(0).equalTo(0).apply((first, second) =>{
            if (second == null) {
                (first._1, first._2, "-")
            } else {
                (first._1, first._2, second._2)
            }
        }).print()

        println("==========================")

        data1.rightOuterJoin(data2).where(0).equalTo(0).apply((first, second) =>{
            if (first == null) {
                (second._1, "-", second._2)
            } else {
                (first._1, first._2, second._2)
            }
        }).print()

        println("===========================")

        data1.fullOuterJoin(data2).where(0).equalTo(0).apply((first, second) =>{
            if (first == null) {
                (second._1, "-", second._2)
            } else if (second == null){
                (first._1, first._2, "-")
            } else {
                (first._1, first._2, second._2)
            }
        }).print()
    }

    def joinFunction(env: ExecutionEnvironment): Unit = {
        val info1 = ListBuffer[(Int, String)]()
        info1.append((1, "hadoop"))
        info1.append((2, "spark"))
        info1.append((3, "flink"))
        info1.append((4, "java"))
        info1.append((5, "Spring"))
        info1.append((6, "Linux"))

        val info2 = ListBuffer[(Int, String)]()
        info2.append((1, "beijing"))
        info2.append((2, "shanghai"))
        info2.append((3, "chengdu"))
        info2.append((4, "guangzhou"))
        info2.append((5, "hangzhou"))
        info2.append((7, "shenzhen"))

        import org.apache.flink.api.scala._
        val data1 = env.fromCollection(info1)
        val data2 = env.fromCollection(info2)

        data1.join(data2).where(0).equalTo(0).apply((first, second) =>{
            (first._1, first._2, second._2)
        }).print()
    }

    def distinctFunction(env: ExecutionEnvironment): Unit = {
        val info = ListBuffer[String]()
        info.append("hadoop,spark")
        info.append("hadoop,flink")
        info.append("flink,flink")
        info.append("flink,flink")

        import org.apache.flink.api.scala._
        val data = env.fromCollection(info)
        data.flatMap(_.split(",")).distinct().print()
    }

    def flatMapFunction(env: ExecutionEnvironment): Unit = {
        val info = ListBuffer[String]()
        info.append("hadoop,spark")
        info.append("hadoop,flink")
        info.append("flink,flink")
        info.append("flink,flink")
        import org.apache.flink.api.scala._
        val data = env.fromCollection(info)
        //data.map(_.split(",")).print()
        //data.flatMap(_.split(",")).print()
        data.flatMap(_.split(",")).map((_, 1)).groupBy(0).sum(1).print()
    }

    def firstFunction(env:ExecutionEnvironment): Unit ={
        val info = ListBuffer[(Int, String)]()
        info.append((1, "hadoop"))
        info.append((1, "spark"))
        info.append((1, "flink"))
        info.append((2, "java"))
        info.append((2, "Spring"))
        info.append((3, "Linux"))
        info.append((4, "Vue"))

        import org.apache.flink.api.scala._
        val data = env.fromCollection(info)
        //data.first(4).print()
        //println("============")
        //data.groupBy(0).first(2).print()
        data.groupBy(0).sortGroup(1, Order.ASCENDING).first(2).print()
    }

    def mapPartitionFunction(env: ExecutionEnvironment): Unit = {
        val students = new ListBuffer[String]
        for (i <- 1 to 100) {
            students.append("student: " + i)
        }

        import org.apache.flink.api.scala._
        val data = env.fromCollection(students).setParallelism(4)
//        data.map(x => {
//            //获取Connection
//            val connection = DBUtils.getConnnection()
//            println(connection + "...")
//
//            //保存数据到DB
//            DBUtils.returnConnection(connection)
//        }).print()
        data.mapPartition(x => {
            val connection = DBUtils.getConnnection()
            println(connection + "....")
            DBUtils.returnConnection(connection)
            x
        }).print()
    }

    def filterFunction(env: ExecutionEnvironment): Unit = {
        import org.apache.flink.api.scala._
        //链式编程
        env.fromCollection(List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10))
            .map(_ + 1)
            .filter(_ > 5)
            .print()
    }

    def mapFunction(env: ExecutionEnvironment): Unit = {
        import org.apache.flink.api.scala._
        val data = env.fromCollection(List(1, 2, 3, 4, 5, 6, 7, 8, 9, 10))
        data.print()
        println("--------------")
        //data.map((x: Int) => x * 2).print()
        //data.map((x) => x * 2).print()
        //data.map(x => x * 2).print()
        data.map(_*2).print()
    }
}
